icu4j/main/core/src/main/java/com/ibm/icu/util/IslamicCalendar.java

```diff
@@ -319,6 +320,294 @@ public IslamicCalendar(int year, int month, int date, int hour,
         this.set(Calendar.SECOND, second);
     }
 
+    // Private interface for different Islamic calendar algorithms.
+    private interface Algorithm {
+        /**
+         * Returns <code>true</code> if this object is using the fixed-cycle civil
+         * calendar, or <code>false</code> if using the religious, astronomical
+         * calendar.
+         */
+        public boolean isCivil();
+
+        /**
+         * Return the type the algorithm implement.
+         */
+        public CalculationType getType();
+
+        /**
+         * Return the epoch used by this algorithm.
+         */
+        public long epoch();
+
+        /**
+         * Return the day # on which the given year starts. Days are counted
+         * from the Hijri epoch, origin 0.
+         *
+         * @param year  The hijri year
+         */
+        public long yearStart(int year);
+
+        /**
+         * Return the day # on which the given month starts. Days are counted
+         * from the Hijri epoch, origin 0.
+         *
+         * @param year  The hijri year
+         * @param month  The hijri month, 0-based
+         */
+        public long monthStart(int year, int month);
+
+        /**
+         * Return the length (in days) of the given month.
+         *
+         * @param extendedYear  The hijri year
+         * @param month The hijri month, 0-based
+         */
+        public int monthLength(int extendedYear, int month);
+
+        /**
+         * Return the length (in days) of the given year.
+         *
+         * @param extendedYear  The hijri year
+         */
+        public int yearLength(int extendedYear);
+
+        /**
+         * Compute the year, month, dayOfMonth, and dayOfYear of the given julian days
+         * and current time. Return an array of 4 integer which contains the resolved
+         * year, month, dayOfMonth and dayOfYear.
+         * @param julianDays
+         * @param current the time in millisecond.
+         */
+        public int[] compute(long julianDays, long current);
```

Comment:
I'm not super bananas on having this function return an array of `int`, but I think I agree it's not really worth the trouble to define another class just to use as the return value of this function.

Reply:
ok, I change to passing 4 IntConsumer to consume the calculated result.